substrate/frame/support/src/traits/tokens/fungible/regular.rs

```diff
@@ -423,19 +420,20 @@ pub trait Balanced<AccountId>: Inspect<AccountId> + Unbalanced<AccountId> {
 	/// This is just the same as burning and issuing the same amount and has no effect on the
 	/// total issuance.
 	///
-	/// This is infallible, but doesn't guarantee that the entire `amount` is used to create the
-	/// pair, for example in the case where the amounts would cause overflow or underflow in
-	/// [`Balanced::issue`] or [`Balanced::rescind`].
-	fn pair(amount: Self::Balance) -> (Debt<AccountId, Self>, Credit<AccountId, Self>) {
-		// Maximum amount which can be issued.
-		let issue_cap = Self::Balance::max_value().saturating_sub(Self::total_issuance());
-		// Maximum amount which can be rescinded.
-		let rescind_cap = Self::total_issuance();
-
-		// Amount which can be both issued and rescinded.
-		let capped_amount = amount.min(issue_cap).min(rescind_cap);
-
-		(Self::rescind(capped_amount), Self::issue(capped_amount))
+	/// This could fail when we cannot issue and redeem the entire `amount`, for example in the
+	/// case where the amount would cause overflow or underflow in [`Balanced::issue`] or
+	/// [`Balanced::rescind`].
+	fn pair(
+		amount: Self::Balance,
+	) -> Result<(Debt<AccountId, Self>, Credit<AccountId, Self>), DispatchError> {
+		let issued = Self::issue(amount);
+		let rescinded = Self::rescind(amount);
+		if issued.peek() != rescinded.peek() {
```

Comment:
it does not guarantee the entire amount is issued 

https://github.com/paritytech/polkadot-sdk/pull/1296/files/3d14f3d0a3dd3f807402f598602000795b01a8d5..c50c714c76cbbd2e3aacff8da4860def7fd5f794#diff-e04f7799b91b4225b1923b49d8484495f7868868ab32ed78919f310ffe268fd9L410-L411

Reply:
Then we should also check and error, or not?

Reply:
Hmm, yeah I guess there is an edge case I missed. 

Here's how the issue/redeem caps I believe are computed (let me know if I'm missing some other ways they could be capped):

```rust
// Maximum amount which can be issued.
let issue_cap = Self::Balance::max_value().saturating_sub(Self::total_issuance());
// Maximum amount which can be rescinded.
let rescind_cap = Self::total_issuance();
```

So if Balance::max_value - total_issuance == total_issuance it could issue below the cap. I'll add another check to make sure the amounts also equal the intended amount.

Reply:
Fixed